### ISSUE TITLE
mlocate incorrectly warns about lack of user support

### ISSUE THREAD

**Issue by unode** [NixOS/nixpkgs#30864]
## Issue description

`nixos-rebuild switch` prints `trace: warning: mlocate does not support searching as user other than root`

The message seems misleading as everything works as expected, even as a regular user.

### Steps to reproduce
    services.locate = {
      enable = true;
      locate = pkgs.mlocate;
    };

## Technical details

* System: (NixOS: `17.09.git.e1b01a0 (Hummingbird)`)
* Nix version: (`nix-env (Nix) 1.11.15`)
* Nixpkgs version: (`17.09.1547.2231575a7d`)

**Comment by ar1a** [comment]
This still happens

**Comment by unode** [comment]
That's surprising since it works for me without root.

I only experienced issues when the wrong `locate` binary is used. If I recall correctly `findutils` has higher priority and causes this.

Can you `which locate` and paste whatever error you get?

**Comment by unode** [comment]
Sorry hit the wrong button by accident.

**Comment by cizra** [comment]
```
$ which locate
/run/wrappers/bin/locate
$ file /run/wrappers/bin/locate
/run/wrappers/bin/locate: setgid executable, regular file, no read permission
$ ls -lh /run/wrappers/bin/locate
-r-x--s--x 1 root mlocate 18K May 16 13:00 /run/wrappers/bin/locate
```

**Comment by andersk** [comment]
The warning is correct, but the wording is misleading. What it’s trying to warn you is not that `locate` will only run as `root`, but rather that `updatedb` will run as `root`.

You can prevent the warning from being shown (with no other effect) by explicitly requesting that `updatedb` run as `root` with `services.locate.localuser = null`.

**Comment by unode** [comment]
> The warning is correct, but the wording is misleading. What it’s trying to warn you is not that `locate` will only run as `root`, but rather that `updatedb` will run as `root`.

This is warning is still confusing to me. If I'm building an index of files, I'd expect that if I run it as a non privileged user many files will not be reachable and therefore these won't end up in the locate database/index. Given the way `locate` works, I'd also expect to have one database per system, not one per user.
So from this my expectation is that `updatedb` can only successfully index *everything* if run as `root`.

In other words, I'm still not sure what is the purpose or goal of the message.
* Is it just that the `updatedb` command *needs* to run as `root`?
  * is this bad in any way?
  * why is this special given the purpose of the tool?
* Is it that the fact that it runs as `root` is in some way a security concern?
* Is this to clarify that despite the existence of an `mlocate` group (and user?) these are not actually used for indexing,

or is the point something else entirely.

Do other `locate` tools work in such a different way that justifies the existence of this warning for `mlocate`?

**Comment by andersk** [comment]
mlocate’s `updatedb` doesn’t have a `--localuser` option like findutils’ `updatedb` does. Therefore, the NixOS module *ignores* `services.locate.localuser` when using mlocate (#26191). It prints a warning to indicate that your expectation has been violated.

The `mlocate` group is there so that the mlocate `locate` binary can be setgid `mlocate` rather than setuid `root`.

Given that mlocate is designed with an `updatedb` that indexes everything and a `locate` that only shows you your own files, there doesn’t seem to be much point in running `updatedb` as another user. If you want that, why not switch back to findutils?

**Comment by unode** [comment]
> mlocate’s `updatedb` doesn’t have a `--localuser` option like findutils’ `updatedb` does. Therefore, the NixOS module _ignores_ `services.locate.localuser` when using mlocate (#26191). It prints a warning to indicate that your expectation has been violated.

I see, in this context, the above makes sense.



> Given that mlocate is designed with an `updatedb` that indexes everything and a `locate` that only shows you your own files, there doesn’t seem to be much point in running mlocate’s `updatedb` as another user. If you want that, why not switch back to findutils?

I actually want the full system indexing. To me `locate` has always been a synonym of `find / ...`, except several times faster.

**Comment by unode** [comment]
@andersk Could you include your `db28ce3 ` commit in a pull request? I think that's a good solution and we can close this issue.

**Comment by andersk** [comment]
@unode It’s been in #79239 this whole time.

**Comment by unode** [comment]
My bad, I read only the latest updates and the pull request was higher up.
Did a bit of noise on the pull request. It's a harmless change so I was surprised to see it sitting there for this long.

**Comment by correabuscar** [comment]
To avoid ```updatedb: can not find group `mlocate'```, looks like both of these are needed:
```
services.locate.package = pkgs.mlocate;
services.locate.localuser = null;
```
when using:
```
environment.systemPackages = with pkgs; [
...
  mlocate
];
```
with or without:
```
services.locate.enable = true;
```

But if not using the `environment.systemPackages = with pkgs; [ mlocate ];` (so only using the `services.locate.enable = true;` alone) then there's no ```updatedb: can not find group `mlocate'``` error.

If anyone wants to guess why, here's the source: https://github.com/NixOS/nixpkgs/blob/nixos-23.11/nixos/modules/misc/locate.nix
